### PR TITLE
pretty big refactor of boll

### PR DESCRIPTION
**Pull request by jcreamer898**
Hey everyone! Long time listener, first time caller.

I see a lot of value in having boll available as a tool, so I've gone in and made some pretty big improvements.

* added `lage` as the `build` and `test` runner.
* upgrade to latest Typescript (^4.9.2)
* add a new API for `addRule`

```js
addRule({
  name: "YourRule",
  check: async (filename, { logger, options }) => {
    // ...  
  }
})
```

* Added a new finder called `WorkspacesGlob` which uses `workspace-tools` under the hood

```js
const MonorepoConfig: ConfigDefinition = {
  name: "boll:recommended/monorepo",
  ruleSets: [
    {
      fileLocator: new WorkspacesGlob(),
      checks: {
        meta: [{ rule: "EnsureBoll" }]
      }
    },
  ]
};
```

Using `WorkspacesGlob` along with the `meta` check will allow users to receive all of a monorepo's `package.json` files in the `check` function as the `files` argument.

* added support for loading boll plugins by specifying `boll:{plugin}` e.g. 

```js
module.exports = {
  extends: "boll:recommended/monorepo"
}
```
Plugins can now simply implement a `bootstrap` function, which will automatically load, and register itself.

The other idea here was to have rules that are at the _repo_ level, not just in the individual packages.

Therefore I can run `boll` at the top level of my monorepo like...

```json
{
  "name": "my-monorepo",
  "scripts": {
    "lint:repo": "boll run"
  },
  "workspaces": {
    "packages": ["packages/*"]  
  } 
}
```

I have a few more potential things to add such as:
* allowing multiple `extends` which I might throw in too, but here's the start. :)
* allowing rules to have a `fix` function

### CODE REVIEW

File: lage.config.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  "pipeline": {
+    "build": ["^build"],
+    "test": ["build"]
```
Comment (kenotron) [inline lage.config.js:4]: Did you need to build before testing, really?

File: package.json [hunk 2]
```diff
@@ -2,23 +2,26 @@
   "author": "JD Huntington <jdh@microsoft.com>",
   "devDependencies": {
     "beachball": "^1.36.2",
-    "lerna": "^3.22.1"
+    "lerna": "^3.22.1",
```
Comment (kenotron) [inline package.json:5]: Do you still need lerna?

File: packages/rules-monorepo/src/ensure-boll.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  check: async (files: FileContext[], options: RuleOptions) => {
+    const errors: { formattedMessage: string; status: number }[] = [];
+
+    await Promise.all(
+      files.map(async ({ filename, content }) => {
+        const pkg = JSON.parse(content);
+
+        let hasBoll = false;
+        let hasIgnore = false;
+        let bollConfig = join(dirname(filename), ".boll.config.js");
```
Comment (kenotron) [inline packages/rules-monorepo/src/ensure-boll.ts:35]: is .boll.config.js the only config name you can use?

**Review by kenotron — COMMENTED** [review]
Making comments here - but in general looks really good!

**Review by kenotron — APPROVED** [review]

File: lage.config.js (hunk 1 above)
Comment (jcreamer898) [reply lage.config.js:4]: In this repo, yeah, haha. I can fix it I guess, but that's just how y'all have this setup.

File: package.json (hunk 2 above)
Comment (jcreamer898) [reply package.json:5]: I'll refactor it away completely. Good call.

File: packages/rules-monorepo/src/ensure-boll.ts (hunk 3 above)
Comment (jcreamer898) [reply packages/rules-monorepo/src/ensure-boll.ts:35]: AFAIK, I'll check tho.

**Pull request merged**